### PR TITLE
[Streaming Generator] Remove busy waiting

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR removes the busy wait for `next(ray.StreamingObjectRefGenerator)`. To do this, we add a sentinel ObjectRef to the end of the streaming generator. This ObjectRef contains a dummy value but is never returned directly to the caller. On `next`, we:
- Compute the next expected ObjectRef from the generator
- Use `ray.wait` to check when the next expected ObjectRef is ready
- If the ObjectRef is the end-of-stream, we raise StopIteration (or return the contained application-level exception) instead of returning the ObjectRef

This also fixes a bug where the first execution fails before the end-of-stream, and then the second execution fails earlier due to nondeterminism. In this case, we mark the first execution's last ObjectRef as the end-of-stream.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -284,145 +282,163 @@ class StreamingObjectRefGenerator:
...
         except ObjectRefStreamEneOfStreamError:
-            raise StopIteration
+            if self._generator_task_exception:
+                # Exception has been returned. raise StopIteration.
+                raise StopIteration
 
-    def _handle_next(self):
-        """Get the next item from the ObjectRefStream.
+            try:
+                ray.get(ref)
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:310]: Never mind, this is only done if we're at the end of the stream, so it should not incur a remote fetch.

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -284,145 +282,163 @@ class StreamingObjectRefGenerator:
...
-        the ref is supposed to be return.
-        """
-        if self._generator_task_exception:
-            # The generator task has failed already.
-            # We raise StopIteration
-            # to conform the next interface in Python.
-            if is_async:
+        ref = core_worker.peek_object_ref_stream(
+            self._generator_ref)
+        ready, unready = await asyncio.wait([ref], timeout=timeout_s)
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:333]: Left a TODO here, FYI @rkooo567. I'm not totally sure if `await` fetches the object local.

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -284,145 +282,77 @@ class StreamingObjectRefGenerator:
...
+            [ref], timeout=timeout_s, fetch_local=False)
+        if len(unready) > 0:
+            return ObjectRef.nil()
 
-    def _handle_next_sync(self):
         try:
-            return self._handle_next()
-        except ObjectRefStreamEneOfStreamError:
-            raise StopIteration
+            ref = core_worker.try_read_next_object_ref_stream(
```
Comment (jjyao) [inline python/ray/_raylet.pyx:301]: Why do we reset the ref, don't we already have the ref from peek_object_ref_stream?

Comment (jjyao) [reply python/ray/_raylet.pyx:301]: Oh, you just use it to see if this is the last one. Can we add a check to make sure the previous ref is the same as this one.

File: python/ray/_raylet.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjyao) [reply python/ray/_raylet.pyx:310]: Yea, I think this is just to get the sentinel ObjectRef

File: src/ray/core_worker/task_manager.cc [hunk 4]
```diff
@@ -137,11 +131,33 @@ bool ObjectRefStream::InsertToStream(const ObjectID &object_id, int64_t item_ind
...
+  end_of_stream_index_ = std::max(max_index_seen_ + 1, item_index);
+
+  auto end_of_stream_id = GetObjectRefAtIndex(end_of_stream_index_);
+  item_index_to_refs_.emplace(end_of_stream_index_, end_of_stream_id);
+  *object_id_in_last_index = end_of_stream_id;
+  return;
+}
+
+ObjectID ObjectRefStream::GetObjectRefAtIndex(int64_t generator_index) const {
+  RAY_CHECK(generator_index < 100 * 1000 * 1000);
```
Comment (jjyao) [inline src/ray/core_worker/task_manager.cc:158]: Could you use the constant defined?

File: src/ray/core_worker/task_manager.cc [hunk 5]
```diff
@@ -137,11 +131,33 @@ bool ObjectRefStream::InsertToStream(const ObjectID &object_id, int64_t item_ind
...
+  // 2 for a single return from a generator task.
```
Comment (jjyao) [inline src/ray/core_worker/task_manager.cc:159]: // index 1 is reserved for the first task return from a generator task itself.

Copied from `TaskSpecification::StreamingGeneratorReturnId`

**Review by jjyao — COMMENTED** [review]
Will continue

File: python/ray/_raylet.pyx (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/_raylet.pyx:334]: As far as I understand, fetch_local=False is not supported for the async actors. I think we should fix the issue as a whole as a follow-up.

**Comment by rkooo567** [comment]
cc @jjyao for a review!

File: src/ray/core_worker/task_manager.cc [hunk 6]
```diff
@@ -107,8 +85,11 @@ bool ObjectRefStream::TemporarilyInsertToStreamIfNeeded(const ObjectID &object_i
 }
 
 bool ObjectRefStream::InsertToStream(const ObjectID &object_id, int64_t item_index) {
-  if (end_of_stream_index_ != -1) {
+  RAY_CHECK_EQ(object_id, GetObjectRefAtIndex(item_index));
+  if (end_of_stream_index_ != -1 && item_index >= end_of_stream_index_) {
     RAY_CHECK(next_index_ <= end_of_stream_index_);
+    // Ignore the index after the end of the stream index.
```
Comment (jjyao) [inline src/ray/core_worker/task_manager.cc:91]: Could you add a comment saying when this can happen?

**Review by jjyao — APPROVED** [review]

**Pull request merged**